File: packages/wasmvm/wasmlib/ts/wasmlib/wasmtypes/scstring.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {WasmDecoder, WasmEncoder} from './codec';
import {Proxy} from './proxy';
import {uint16Decode, uint16Encode} from "./scuint16";

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function stringDecode(dec: WasmDecoder): string {
    const length = uint16Decode(dec);
    return stringFromBytes(dec.fixedBytes(length as u32));
}

export function stringEncode(enc: WasmEncoder, value: string): void {
    const buf = stringToBytes(value);
    uint16Encode(enc, buf.length as u16);
    enc.fixedBytes(buf, buf.length as u32);
}

export function stringFromBytes(buf: Uint8Array): string {
    return new TextDecoder().decode(buf);
}

export function stringToBytes(value: string): Uint8Array {
    return new TextEncoder().encode(value);
}

export function stringFromString(value: string): string {
    return value;
}

export function stringToString(value: string): string {
    return value;
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableString {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return this.value();
    }

    value(): string {
        return stringFromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableString extends ScImmutableString {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: string): void {
        this.proxy.set(stringToBytes(value));
    }
}

File: packages/wasmvm/wasmlib/go/wasmlib/wasmtypes/scstring.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmtypes

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

func StringDecode(dec *WasmDecoder) string {
	return string(dec.Bytes())
}

func StringEncode(enc *WasmEncoder, value string) {
	enc.Bytes([]byte(value))
}

func StringFromBytes(buf []byte) string {
	return string(buf)
}

func StringToBytes(value string) []byte {
	return []byte(value)
}

func StringToString(value string) string {
	return value
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScImmutableString struct {
	proxy Proxy
}

func NewScImmutableString(proxy Proxy) ScImmutableString {
	return ScImmutableString{proxy: proxy}
}

func (o ScImmutableString) Exists() bool {
	return o.proxy.Exists()
}

func (o ScImmutableString) String() string {
	return StringToString(o.Value())
}

func (o ScImmutableString) Value() string {
	return StringFromBytes(o.proxy.Get())
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScMutableString struct {
	ScImmutableString
}

func NewScMutableString(proxy Proxy) ScMutableString {
	return ScMutableString{ScImmutableString{proxy: proxy}}
}

func (o ScMutableString) Delete() {
	o.proxy.Delete()
}

func (o ScMutableString) SetValue(value string) {
	o.proxy.Set(StringToBytes(value))
}
